frontend/public/components/monitoring/query-browser.tsx

```diff
@@ -385,8 +378,8 @@ const Graph: React.FC<GraphProps> = React.memo(
       // Set a reasonable Y-axis range based on the min and max values in the data
       const findMin = (series: GraphSeries) => _.minBy(series, 'y');
       const findMax = (series: GraphSeries) => _.maxBy(series, 'y');
-      let minY: number = findMin(data.map(findMin))?.y ?? 0;
-      let maxY: number = findMax(data.map(findMax))?.y ?? 0;
+      let minY: number = findMin((data as GraphSeries[]).map(findMin))?.y ?? 0;
+      let maxY: number = findMax((data as GraphSeries[]).map(findMax))?.y ?? 0;
```

Comment:
Honestly not sure why TypeScript 4 doesn't seem to correctly infer this type, but I felt there was little harm in adding the `as`. This change is just for upgrading TypeScript, so I could remove it from this PR if you prefer?

Reply:
Type assertions like this can actually cause bugs since TypeScript will trust what you say even if it's not true. It's usually an indication that there is a missing or incorrect type declaration somewhere, and it's better to fix that.

https://medium.com/swlh/why-you-should-avoid-type-assertions-in-typescript-5494e3d04dd
https://dev.to/shakyshane/type-assertions-vs-declarations-the-dangers-of-as-in-typescript-1caf

Reply:
Looking at the code, I don't know why TypeScript doesn't know the correct type either :/

Reply:
Thanks for the additional explanation and for looking at the code @spadgett!

The problem seems to be because TypeScript gets confused by this
```
if (data.every(_.isEmpty)) {
  return <GraphEmpty />;
}
```

After this code, TypeScript thinks that `data` is `never`.

I changed it to `if (!data.some(Array.isArray)) { ... }`, which TypeScript interprets correctly. I think using the JavaScript native `Array.isArray` is probably preferable to `_.isEmpty` anyway.